airbyte-cdk/python/airbyte_cdk/sources/file_based/stream/default_file_based_stream.py

```diff
@@ -152,17 +152,18 @@ def _get_raw_json_schema(self) -> JsonSchema:
         elif self.config.schemaless:
             return schemaless_schema
         else:
-            files = self.list_files()
-            total_n_files = len(files)
+            all_files = self.list_files()
+            files_to_read = list(self._cursor.get_files_to_sync(all_files, self.logger))
```

Comment:
This part is worth noting. The existing schema inference code never relied on the cursor (although it is defined by default). This modifies the flow a bit to also apply the cursor to filter out records for both `discover` and `read`.

A few alternative options I explored:
- Using `start_date` in `AbstractFileBasedStreamReader.filter_files_by_globs()`. This has the nice benefit of always being applied even if the source implements their own FileBasedCursor. But it has the drawback of inserting more modified comparisons across the codebase which I don't love
- We also could just add a standalone check here to filter out files here or in `self.list_files()`, but the datetime checks are meant to be done in the cursor.

Up for discussion, but I think incorporating it here feels sensible since we do plan on a stateful discover at some point which will need the cursor.

Reply:
A couple of thoughts -

I think I'd prefer not to restrict schema inference to only files that are supposed to be synced, because it will potentially be a totally normal case for there to just be 0 or 1 files that need to be synced, in which case schema inference will throw an exception (for the 0 files case) or only infer from the one file, which is more error prone.

That said, it makes sense to me to not include files from before `start_date` in schema inference.

So it's feeling to me like `filter_files_by_globs` would be a good place for it (although if we do that it we'd want to rename it). Also worth noting, if no files are returned from `filter_files_by_globs`, we currently fail `check`, because the implication is that we can't do schema inference and that there's a possible misconfiguration. If we incorporate a `start_date` check in there and there are no files before `start_date`, the failure might be useful to the user (i.e. don't have to wait until a discover/sync to see that their stream is empty).

Reply:
That's makes sense! That was the original approach that we had discussed, but this redesign is probably over-consolidating how we decide what files to pull from the file store. I will incorporate the feedback and revert back to old approach where we use `filter_files_by_globs_and_start_date()`

Reply:
Moving to `filter_files_by_globs` makes sense to me. The function would effectively become a filter on the name and timestamp, which seems reasonable.